File: BedivereKnx/EventArguments.vb
Imports Knx.Falcon

Public Delegate Sub GroupReadHandler(e As KnxReadEventArgs)
Public Delegate Sub GroupWriteHandler(e As KnxWriteEventArgs)
Public Delegate Sub KnxMessageHandler(e As KnxMsgEventArgs, log As String)
Public Delegate Sub ScheduleEventHandler(code As String, e As KnxWriteEventArgs)
Public Delegate Sub ScheduleTimerHandler(e As KnxScheduleTimerState)
Public Delegate Sub KnxDeviceStateHandler(sender As KnxDeviceInfo, e As IndAddressState)

Public Class KnxReadEventArgs

    Inherits EventArgs

    Public ReadOnly Property InterfaceCode As String

    Public ReadOnly Property GroupAddr As GroupAddress

    Public ReadOnly Property Priority As MessagePriority = MessagePriority.Low

    Public Sub New(IfCode As String, GrpAddr As GroupAddress, Optional Priority As MessagePriority = MessagePriority.Low)
        _InterfaceCode = IfCode
        _GroupAddr = GrpAddr
        _Priority = Priority
    End Sub

End Class

Public Class KnxWriteEventArgs

    Inherits EventArgs

    Public ReadOnly Property InterfaceCode As String

    Public ReadOnly Property GroupAddr As GroupAddress

    Public ReadOnly Property GroupVal As GroupValue

    Public ReadOnly Property Priority As MessagePriority = MessagePriority.Low

    Public Sub New(IfCode As String, GrpAddr As GroupAddress, GrpVal As GroupValue, Optional Priority As MessagePriority = MessagePriority.Low)
        _InterfaceCode = IfCode
        _GroupAddr = GrpAddr
        _GroupVal = GrpVal
        _Priority = Priority
    End Sub

End Class

Public Class KnxMsgEventArgs

    Inherits EventArgs

    Public ReadOnly Property MessageType As KnxMessageType

    Public ReadOnly Property EventType As GroupEventType

    Public ReadOnly Property MessagePriority As MessagePriority

    Public ReadOnly Property HopCount As Byte

    Public ReadOnly Property DestinationAddress As GroupAddress

    Public ReadOnly Property SourceAddress As IndividualAddress

    Public ReadOnly Property IsSecure As Boolean

    Public ReadOnly Property Value As GroupValue

    Public Sub New(MsgType As KnxMessageType, EvtType As GroupEventType, MsgPriority As MessagePriority, HopCount As Byte, DesAddr As GroupAddress, SourceAddr As IndividualAddress, Optional IsSecure As Boolean = False, Optional Val As GroupValue = Nothing)
        _MessageType = MsgType
        _EventType = EvtType
        _MessagePriority = MsgPriority
        _HopCount = HopCount
        _DestinationAddress = DesAddr
        _SourceAddress = SourceAddr
        _IsSecure = IsSecure
        _Value = Val
    End Sub

    Public Sub New(MsgType As KnxMessageType, GrpEvtArgs As GroupEventArgs)
        _MessageType = MsgType
        _EventType = GrpEvtArgs.EventType
        _MessagePriority = GrpEvtArgs.MessagePriority
        _HopCount = GrpEvtArgs.HopCount
        _DestinationAddress = GrpEvtArgs.DestinationAddress
        _SourceAddress = GrpEvtArgs.SourceAddress
        _IsSecure = GrpEvtArgs.IsSecure
        _Value = GrpEvtArgs.Value
    End Sub

End Class

File: BedivereKnx/EventArguments.cs
using BedivereKnx.KnxSystem;
using Knx.Falcon;

namespace BedivereKnx
{

    /// <summary>
    /// 值变化委托
    /// </summary>
    public delegate void ValueChangeHandler<T>(T? value);

    /// <summary>
    /// KNX组读取委托
    /// </summary>
    /// <param name="e"></param>
    public delegate void GroupReadHandler(KnxGroupEventArgs e);

    /// <summary>
    /// KNX组写入委托
    /// </summary>
    public delegate void GroupWriteHandler(KnxGroupEventArgs e, GroupValue value);

    /// <summary>
    /// KNX报文传输委托
    /// </summary>
    public delegate void KnxMessageHandler(KnxMsgEventArgs e, string? log);

    /// <summary>
    /// KNX设备状态委托
    /// </summary>
    /// <param name="dev"></param>
    /// <param name="state"></param>
    public delegate void KnxDeviceStateHandler(KnxDeviceInfo dev, KnxDeviceState state);

    /// <summary>
    /// 时间表事件委托
    /// </summary>
    /// <param name="code"></param>
    /// <param name="e"></param>
    /// <param name="value"></param>
    public delegate void ScheduleEventHandler(string log, KnxGroupEventArgs e, GroupValue value);

    /// <summary>
    /// 时间表定时器委托
    /// </summary>
    /// <param name="state"></param>
    public delegate void ScheduleTimerHandler(KnxScheduleTimerState state);

    public class KnxGroupEventArgs(string? interfaceCode, GroupAddress groupAddress, MessagePriority priority = MessagePriority.Low) : EventArgs
    {

        public string? InterfaceCode { get; } = interfaceCode;

        public GroupAddress GroupAddress { get; } = groupAddress;

        public MessagePriority Priority { get; } = priority;

    }

    public class KnxMsgEventArgs : EventArgs
    {
        public KnxMessageType MessageType { get; }

        public GroupEventType EventType { get; }

        public MessagePriority MessagePriority { get; }

        public byte HopCount { get; }

        public GroupAddress DestinationAddress { get; }

        public IndividualAddress SourceAddress { get; }

        public bool IsSecure { get; }

        public GroupValue? Value { get; }

        public KnxMsgEventArgs(KnxMessageType type, GroupEventArgs arg)
        {
            MessageType = type;
            EventType = arg.EventType;
            MessagePriority = arg.MessagePriority;
            HopCount = arg.HopCount;
            DestinationAddress = arg.DestinationAddress;
            SourceAddress = arg.SourceAddress;
            IsSecure = arg.IsSecure;
            Value = arg.Value;
        }

        public KnxMsgEventArgs(KnxMessageType msgType,
                               GroupEventType eventType,
                               MessagePriority priority,
                               byte hop,
                               GroupAddress grpAddr,
                               IndividualAddress indAddr,
                               bool isSecure = false,
                               GroupValue? val = null)
        {
            MessageType = msgType;
            EventType = eventType;
            MessagePriority = priority;
            HopCount = hop;
            DestinationAddress = grpAddr;
            SourceAddress = indAddr;
            IsSecure = isSecure;
            Value = val;
        }

    }

}
